### PR TITLE
DEV-4043: Payment model

### PR DESCRIPTION
**Pull request by benjaminEwhite**
Please fill out each section even if it's just with "N/A"

#### What's this PR do?

This PR introduces the per-Stripe-transaction `Payment` model which we'll use in future release to represent both successful charges and successful refunds, tying those back to NRE contributions.

In addition to the model definition and minimal changes to provide a read-only admin view, the Payment model contains three class methods that correspond to the three types of webhooks we'll listen for to create payments starting in [DEV-4045](https://news-revenue-hub.atlassian.net/browse/DEV-4045):

- `from_stripe_payment_intent_succeeded_event`
- `from_stripe_charge_refunded_event`
- `from_stripe_invoice_payment_succeeded_event`

There is a fairly meticulous testing strategy around these methods involving fixture data harvested from actual Stripe events and API calls (note this data is sanitized). This was put in scope in this ticket to ensure the data model definition would accommodate achievable workflows.

**RE: changes to settings.dev.py**

We used to have dev.py dynamically change the Django cache settings if the broker url was for in-memory setup.  In deploy.py, we use Redis cache, so use case for dev.py has been running tests locally (via import of dev.py in local.py) and in CI.

In CI, we're not doing anything to manually set BROKER_URL, which means it should be defaulting to `memory://`.  

It was helpful to shift from DummyCache to LocMemCache because it worked better with testing Stripe caching layer around payment instances.  [See here](https://github.com/newsrevenuehub/rev-engine/pull/1291/files#diff-1efa81c75680bae69c0dab2b2f78be3f0c4e06f38441d9d819399e75017cc5f4R1889-R1901).

At present, there doesn't seem to be any use case around conditionally shifting to DummyCache if using memory for broker.

#### Why are we doing this? How does it help us?

- This will allow us to roll up individual payments for recurring contributions into detail view, which is required by both org and contributor users
- Will allow us to provide much better reporting on contributions revenue

#### Are there detailed, specific, step-by-step testing instructions in the associated ticket?

Yes

#### Did this PR make changes to the user interface that may require changes to the user-facing docs?

No

#### Have automated unit tests been added? If not, why?

Yes

#### How should this change be communicated to end users?

N/A

#### Are there any smells or added technical debt to note?

No

#### Has this been documented? If so, where?

No.

#### What are the relevant tickets? Add a link to any relevant ones.

- [DEV-4043](https://news-revenue-hub.atlassian.net/browse/DEV-4043)
- [DEV-4045](https://news-revenue-hub.atlassian.net/browse/DEV-4045)

#### Do any changes need to be made before deployment to production (adding environment variables, for example)? If so, open a ticket and link it to the ticket for this PR and list it here:

No

#### Are there next steps? If so, what? Have tickets been opened for them? List next-step tickets here:

N/A

[DEV-4045]: https://news-revenue-hub.atlassian.net/browse/DEV-4045?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[DEV-4043]: https://news-revenue-hub.atlassian.net/browse/DEV-4043?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: apps/contributions/serializers.py [hunk 1]
```diff
@@ -226,103 +220,6 @@ def to_internal_value(self, data):
         return super().to_internal_value(data)
 
 
-class StripeMetadataSchemaBase(pydantic.BaseModel):
```
Comment (benjaminEwhite) [inline apps/contributions/serializers.py:229]: Moved to apps.contributions.types

File: apps/contributions/tests/test_models.py [hunk 2]
```diff
@@ -1,11 +1,14 @@
 import datetime
+import json
```
Comment (njharman) [inline apps/contributions/tests/test_models.py:2]: I don't think i saw test i was looking for.

One(s) that exercises many to one relationship. Specifically test that adds multiple Payments to recurring Contribution.  [I wanted to ref AC, but the many-to-one is not AC.

File: apps/contributions/models.py [hunk 3]
```diff
@@ -926,3 +937,286 @@ def fix_missing_contribution_metadata(cls, dry_run: bool = False) -> None:
...
+            if event_types and event.type not in event_types:
+                raise ValueError(Payment.EVENT_IS_UNEXPECTED_TYPE_ERROR_MSG_TEMPLATE.format(event_types=event_types))
+            return func(*args, **kwargs)
+
+        return wrapper
+
+    return decorator
+
+
+class Payment(IndexedTimeStampedModel):
```
Comment (njharman) [inline apps/contributions/models.py:971]: surprised there was no requirement for timestamp/date

**Review by njharman — CHANGES_REQUESTED** [review]
Good coverage

File: apps/contributions/models.py (hunk 3 above)
Comment (benjaminEwhite) [reply apps/contributions/models.py:971]: we get created/modified via IndexedTimeStampedModel, or are you looking for something different?

File: apps/contributions/types.py [hunk 4]
```diff
@@ -51,3 +59,114 @@ class Config:
...
+        if len(v) > settings.METADATA_MAX_SWAG_CHOICES_LENGTH:
+            raise ValueError("swag_choices is too long")
+        choices = v.split(cls.SWAG_CHOICES_DELIMITER)
+        # for instance, "tshirt" or "tshirt:hoodie"
+        choice_pattern = rf"[\w-]+({cls.SWAG_SUB_CHOICE_DELIMITER}[\w]+)?"
+        for choice in choices:
+            # we check if choice is truthy to allow for case of a hanging `;` leading to an empty choice
+            if choice and not re.fullmatch(choice_pattern, choice):
+                raise ValueError("swag_choices is not valid")
+        return v
```
[the remark this answers is not in the document]
Comment (benjaminEwhite) [reply apps/contributions/types.py:172]: Correct. It is moved code.

File: apps/contributions/models.py [hunk 5]
```diff
@@ -926,3 +937,286 @@ def fix_missing_contribution_metadata(cls, dry_run: bool = False) -> None:
...
+        cache_key = f"_get_stripe_balance_transaction_{balance_transaction_id}_{account_id}_{'_'.join(expand) if expand else 'no_expand'}"
+        cached_result = cache.get(cache_key)
+        if cached_result is not None:
+            logger.debug("Found cached result for %s", cache_key)
+            return stripe.BalanceTransaction.construct_from(cached_result, key=stripe.api_key)
+
+        logger.debug("No cached result found for %s", cache_key)
+        result = stripe.BalanceTransaction.retrieve(balance_transaction_id, stripe_account=account_id)
+        # load/dump gets us fully serializable data suited for caching
+        cache.set(cache_key, json.loads(json.dumps(result)), 60 * 3)
```
[the remark this answers is not in the document]
Comment (benjaminEwhite) [reply apps/contributions/models.py]: 1c2980127c

File: apps/contributions/models.py [hunk 6]
```diff
@@ -926,3 +937,286 @@ def fix_missing_contribution_metadata(cls, dry_run: bool = False) -> None:
...
+            )
+
+        cls._ensure_pi_has_single_charge(pi, event.id)
+        try:
+            contribution = Contribution.objects.get(provider_payment_id=pi.id)
+        except Contribution.DoesNotExist:
+            try:
+                contribution = (
+                    cls.get_contribution_for_recurrence(balance_transaction_id, event.account)
+                    if balance_transaction
```
[the remark this answers is not in the document]
Comment (benjaminEwhite) [reply apps/contributions/models.py:1110]: 0538c77

File: apps/contributions/models.py [hunk 7]
```diff
@@ -926,3 +937,286 @@ def fix_missing_contribution_metadata(cls, dry_run: bool = False) -> None:
...
+    def _handle_create_payment(
+        cls,
+        contribution: Contribution | None,
+        balance_transaction: stripe.BalanceTransaction | None,
+        amount_refunded: int = 0,
+    ) -> Payment:
+        if not contribution:
+            raise ValueError("Could not find a contribution for this event")
+        if not balance_transaction:
+            raise ValueError("Could not find a balance transaction for this event")
```
[the remark this answers is not in the document]
Comment (benjaminEwhite) [reply apps/contributions/models.py]: 285888b3c8

File: revengine/settings/base.py [hunk 8]
```diff
@@ -248,7 +248,7 @@ class CurrencyDict(TypedDict):
 
 CACHES = {
     "default": {
-        "BACKEND": "django_redis.cache.RedisCache",
+        "BACKEND": "django.core.cache.backends.locmem.LocMemCache",
```
[the remark this answers is not in the document]
Comment (benjaminEwhite) [reply revengine/settings/base.py:253]: yes. see settings/deploy.py:

```
CACHES = {
    "default": {
        "BACKEND": "django_redis.cache.RedisCache",
        "LOCATION": f"{CACHE_HOST}/0",
        "OPTIONS": {
            "CLIENT_CLASS": "django_redis.client.DefaultClient",
            "CONNECTION_POOL_KWARGS": CONNECTION_POOL_KWARGS,
        },
    }
}
```

File: apps/contributions/tests/test_models.py [hunk 9]
```diff
@@ -1709,3 +1719,538 @@ def test_filter_queryset_for_contributor_excludes_statuses_other_than_paid(
...
+            == payment.contribution.donation_page.revenue_program.payment_provider.stripe_account_id
+        )
+
+    @pytest.fixture(autouse=True)
+    def use_mem_cache(self, settings):
+        settings.CACHES = {
+            "default": {
+                "BACKEND": "django.core.cache.backends.locmem.LocMemCache",
+            }
+        }
```
[the remark this answers is not in the document]
Comment (benjaminEwhite) [reply apps/contributions/tests/test_models.py]: 186d966

File: apps/contributions/tests/test_models.py (hunk 2 above)
Comment (benjaminEwhite) [reply apps/contributions/tests/test_models.py:2]: This help? 60b702d

**Review by njharman — APPROVED** [review]

File: apps/contributions/models.py [hunk 10]
```diff
@@ -924,3 +935,284 @@ def fix_missing_contribution_metadata(cls, dry_run: bool = False) -> None:
...
+        balance_transaction: stripe.BalanceTransaction | None,
+        amount_refunded: int = 0,
+        event_id: str = None,
+    ) -> Payment:
+        if not contribution:
+            logger.warning("Cannot find contribution for event %s", event_id)
+            raise ValueError("Could not find a contribution for this event")
+        if not balance_transaction:
+            logger.warning("Cannot find balance transaction for event %s", event_id)
+            raise ValueError("Could not find a balance transaction for this event")
```
Comment (x110dc) [inline apps/contributions/models.py:1181]: Again, why both the warning and the exception?

File: revengine/settings/dev.py [hunk 11]
```diff
@@ -44,9 +44,5 @@
 CELERY_CACHE_BACKEND = BROKER_URL
 CELERY_IMPORTS = ("apps.emails.tasks",)
 
-if BROKER_URL.startswith("memory"):
```
Comment (x110dc) [inline revengine/settings/dev.py:47]: Why was this changed? Should be mentioned in the PR description, I'd think.

**Review by x110dc — CHANGES_REQUESTED** [review]
A few questions.

File: apps/contributions/models.py [hunk 12]
```diff
@@ -924,3 +935,284 @@ def fix_missing_contribution_metadata(cls, dry_run: bool = False) -> None:
...
+    def get_valid_metadata(cls, metadata: dict | None, schema=StripePaymentMetadataSchemaV1_4) -> bool | None:
+        if not metadata:
+            return None
+        try:
+            return schema(**metadata)
+        except PydanticValidationError:
+            return None
+
+    @classmethod
+    def get_subscription_id_for_balance_transaction(
```
[the remark this answers is not in the document]
Comment (benjaminEwhite) [reply apps/contributions/models.py:1034]: No. This returns an ID. I believe I changed implementation along the way and didn't update type def.

File: apps/contributions/models.py [hunk 13]
```diff
@@ -924,3 +935,284 @@ def fix_missing_contribution_metadata(cls, dry_run: bool = False) -> None:
...
+                subscription_id,
+                stripe_account_id,
+            )
+            return None
+
+    @staticmethod
+    def _ensure_pi_has_single_charge(pi: stripe.PaymentIntent, event_id: str) -> None:
+        if not (pi and pi.charges and pi.charges.data and len(pi.charges.data) == 1):
+            logger.warning("Cannot link pi with ID %s for event %s to a single balance transaction", pi.id, event_id)
+            raise ValueError("Cannot link payment intent to a single balance transaction")
```
[the remark this answers is not in the document]
Comment (benjaminEwhite) [reply apps/contributions/models.py:1063]: Seems like a good idea to guarantee a log from this method, and lave it to calling context to decide what to do if an error occurs.

File: apps/contributions/models.py [hunk 14]
```diff
@@ -924,3 +935,284 @@ def fix_missing_contribution_metadata(cls, dry_run: bool = False) -> None:
...
+
+        If there is more than one possible balance transaction (because > 1 charge on PI), we raise an exception. This is not
+        expected, but the data model would allow for it.
+        """
+        # we re-retrieve the PI because its state could have changed between the time the event was received and now
+        pi = stripe.PaymentIntent.retrieve(event.data.object.id, stripe_account=event.account)
+        balance_transaction_id = None
+        if pi and pi.charges and pi.charges.data and len(pi.charges.data) == 1:
+            balance_transaction_id = pi.charges.data[0].balance_transaction
+        cls._ensure_pi_has_single_charge(pi, event.id)
```
[the remark this answers is not in the document]
Comment (benjaminEwhite) [reply apps/contributions/models.py]: And then some. I lightly refactored here: 56ef8344ae

File: apps/contributions/models.py (hunk 10 above)
Comment (benjaminEwhite) [reply apps/contributions/models.py:1182]: Same reason as above. I think it's good for this method to log that something went wrong, and leave itto calling context to decide whether to halt or something different.

File: revengine/settings/dev.py (hunk 11 above)
Comment (benjaminEwhite) [reply revengine/settings/dev.py:47]: Added to PR description.

File: apps/contributions/models.py [hunk 15]
```diff
@@ -924,3 +935,284 @@ def fix_missing_contribution_metadata(cls, dry_run: bool = False) -> None:
...
+            return stripe.BalanceTransaction.construct_from(cached_result, key=stripe.api_key)
+
+        logger.debug("No cached result found for %s", cache_key)
+        result = stripe.BalanceTransaction.retrieve(balance_transaction_id, stripe_account=account_id)
+        # load/dump gets us fully serializable data suited for caching
+        cache.set(cache_key, json.loads(json.dumps(result)), settings.RETRIEVED_STRIPE_ENTITY_CACHE_TTL)
+        return result
+
+    @classmethod
+    def get_valid_metadata(cls, metadata: dict | None, schema=StripePaymentMetadataSchemaV1_4) -> bool | None:
```
[the remark this answers is not in the document]
Comment (benjaminEwhite) [reply apps/contributions/models.py]: This was more broadly in use at one point, but then moved away from it. Removed in daa1beb16e2b

File: apps/contributions/models.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (x110dc) [reply apps/contributions/models.py:1020]: Thanks for fixing the type.

File: apps/contributions/models.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (x110dc) [reply apps/contributions/models.py:1049]: In this case what does the calling context do? Will we have two Sentry alerts for the same issue?

File: apps/contributions/models.py (hunk 10 above)
Comment (x110dc) [reply apps/contributions/models.py:1168]: Same response as above.

File: revengine/settings/dev.py (hunk 11 above)
Comment (x110dc) [reply revengine/settings/dev.py:47]: 👍

**Review by x110dc — APPROVED** [review]
Would be nice to have answer on 2 Sentry alerts but not essential and approved either way.

File: apps/contributions/models.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (benjaminEwhite) [reply apps/contributions/models.py:1048]: 30f762868

File: apps/contributions/models.py (hunk 10 above)
Comment (benjaminEwhite) [reply apps/contributions/models.py:1167]: 30f762868

**Comment by benjaminEwhite** [comment]
@x110dc removed one of the sentry logs in 30f762868

**Pull request merged**